src/datachain/query/dataset.py

```diff
@@ -1383,6 +1383,9 @@ def order_by(self, *args) -> "Self":
     @detach
     def limit(self, n: int) -> "Self":
         query = self.clone(new_table=False)
+        for step in query.steps:
+            if isinstance(step, SQLLimit) and step.n < n:
+                return query
```

Comment:
This will miscompile if the chain contains steps that increase the size of the dataset, like `.gen()`.